xls/modules/fp_sincos_32.x

```diff
@@ -0,0 +1,301 @@
+// guarantee.
+//
+//   The two known misprints in the book are repaired here in the
+// source listings for the gamma function and the incomplete beta
+// integral.
+//
+//   Stephen L. Moshier
+//   moshier@na-net.ornl.gov
+// ====================================================
+
+
+// ====================================================
+//      Circular sine
+//
+// SYNOPSIS:
+//
+// double x, y, sin();
+// y = sin( x );
+//
+// DESCRIPTION:
+//
+// Range reduction is into intervals of pi/4.  The reduction error is nearly
+// eliminated by contriving an extended precision modular arithmetic.
+//
+// Two polynomial approximating functions are employed.
+// Between 0 and pi/4 the sine is approximated by
+//      x  +  x**3 P(x**2).
+// Between pi/4 and pi/2 the cosine is represented as
+//      1  -  x**2 Q(x**2).
+//
+// ACCURACY:
+//  ...
+//
+// Partial loss of accuracy begins to occur at x = 2**30 = 1.074e9.  The loss
+// is not gradual, but jumps suddenly to about 1 part in 10e7.  Results may
+// be meaningless for x > 2**49 = 5.6e14.
+//
+//      cos.c
+//
+//      Circular cosine
+//
+// SYNOPSIS:
+//
+// double x, y, cos();
+// y = cos( x );
+//
+// DESCRIPTION:
+//
+// Range reduction is into intervals of pi/4.  The reduction error is nearly
+// eliminated by contriving an extended precision modular arithmetic.
+//
+// Two polynomial approximating functions are employed.
+// Between 0 and pi/4 the cosine is approximated by
+//      1  -  x**2 Q(x**2).
+// Between pi/4 and pi/2 the sine is represented as
+//      x  +  x**3 P(x**2).
+//
+// ====================================================
+
+// This file implements the cos function for double-recision
+// floating point numbers using a taylor's series approximation.
+// Note:
+//  - Input denormals are treated as/flushed to 0.
+//      (denormals-are-zero / DAZ).
+
+import std
+import float32
+
+import xls.modules.fpadd_2x32
+import xls.modules.fpmul_2x32
+import xls.modules.fpsub_2x32
+import xls.modules.fp_trig_reduce
+
+type F32 = float32::F32;
+
+const SIN_COEF = F32[6]:[
+  float32::unflatten(u32:0x2f2ec7e9),
+  float32::unflatten(u32:0xb2d72f2d),
+  float32::unflatten(u32:0x3638ef1b),
+  float32::unflatten(u32:0xb9500d01),
+  float32::unflatten(u32:0x3c088889),
+  float32::unflatten(u32:0xbe2aaaab),
+];
+
+const COS_COEF = F32[6]:[
+  float32::unflatten(u32:0xad47d24d),
+  float32::unflatten(u32:0x310f74ec),
+  float32::unflatten(u32:0xb493f27c),
+  float32::unflatten(u32:0x37d00d01),
+  float32::unflatten(u32:0xbab60b61),
+  float32::unflatten(u32:0x3d2aaaab),
+];
+
+// Helper function that calculates taylor
+// series approximation of cos in range [0, Pi/4).
+fn cos_taylor(z_sq: F32) -> F32 {
+  const one = float32::one(u1:0);
+  const neg_one = float32::one(u1:1);
+
+  // cos = 1.0 
+  //       - 0.5*z_sq 
+  //       + z_sq*z_sq*((((((_cos[0]*z_sq)+_cos[1])*z_sq+_cos[2])*z_sq+_cos[3])*z_sq+_cos[4])*z_sq+_cos[5])
```

Comment:
nit: can this line be split across several lines and formatted? It's hard to read as is.

Reply:
Done.